src/storage/FileResourceStore.ts

```diff
@@ -0,0 +1,458 @@
+        reject(new ConflictHttpError('Resource with that identifier already exists.'));
+      } catch (error) {
+        // Identifier doesn't exist yet so we can create a container.
+        await this.createContainer(`${this.root}${resultingAction.parentContainer}`,
+          resultingAction.newIdentifier,
+          true,
+          metadata)
+          .then((): any => resolve())
+          .catch((error_): any => {
+            reject(error_);
+          });
+      }
+    });
+  }
+
+  /**
+   * Strips the base from the identifier and checks if it is in the store.
+   * @param identifier - Incoming identifier.
+   *
+   * @throws {@link NotFoundHttpError}
+   * If the identifier does not match the base path of the store.
+   */
+  private parseIdentifier(identifier: ResourceIdentifier): string {
+    const path = identifier.path.slice(this.base.length);
+    if (!identifier.path.startsWith(this.base)) {
+      throw new NotFoundHttpError();
+    }
+    return path;
+  }
+
+  /**
+   * Strips the root path from the filepath and adds the base in front of it.
+   * @param path - The filepath.
+   *
+   * @throws {@Link InternalServerError}
+   * If the filepath does not match the root path of the store.
+   */
+  private mapFilepathToUrl(path: string): string {
+    if (!path.startsWith(this.root)) {
+      throw new InternalServerError('Cannot map filepath to URL.');
+    }
+    return this.base + path.slice(this.root.length);
+  }
+
+  /**
+   * Create a file to represent a resource.
+   * @param path - The path to the directory in which the file should be created.
+   * @param resourceName - The name of the file to be created.
+   * @param data - The data to be put in the file.
+   * @param allowRecursiveCreation - Whether necessary but not existing intermediate containers may be created.
+   * @param metadata - Optional metadata that will be stored at `path/resourceName.metadata` if set.
+   *
+   * @returns Promise of the identifier of the newly created resource.
+   */
+  private async createFile(path: string, resourceName: string, data: Readable,
+    allowRecursiveCreation: boolean, metadata?: Readable): Promise<ResourceIdentifier> {
+    if (allowRecursiveCreation) {
+      await this.createContainer(path, '', true);
+    }
+    return new Promise(async(resolve, reject): Promise<any> => {
+      try {
+        const stats = await fsPromises.lstat(path);
+        if (!stats.isDirectory()) {
+          reject(new MethodNotAllowedHttpError('The given path is not a valid container.'));
+        } else {
+          if (metadata) {
+            this.createMetadata(`${path}${resourceName}.metadata`, metadata, reject);
+          }
+          const writeStream = createWriteStream(path + resourceName);
+
+          data.on('data', (chunk): any => writeStream.write(chunk));
+          data.on('error', reject);
+          data.on('end', (): any => writeStream.end());
```

Comment:
The above block can be simplified to `const writeStream = data.pipe(createWriteStream(...));`, although you probably still need to catch the error on the data stream, should check that.